File: algorithms/[80]删除有序数组中的重复项 II/solution.py
# topics = ["数组", "双指针"]

from typing import List


class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        n = len(nums)
        i = j = 2

        # 当快指针遍历完数组，慢指针指向的为不重复数组结尾 + 1
        while j < n:
            if i < 2 or nums[i - 2] != nums[j]:
                nums[i] = nums[j]
                i += 1
            # 出现第三个相同的元素，快指针单独前进
            j += 1

        return i

File: algorithms/[80]删除有序数组中的重复项 II/solution.ts
// topics = ["数组", "双指针"]

function removeDuplicates(nums: number[]): number {
  const n = nums.length;
  if (n <= 2) {
    return n;
  }
  let slow = 2;
  let fast = 2;
  while (fast < n) {
    if (nums[slow - 2] != nums[fast]) {
      nums[slow] = nums[fast];
      slow += 1;
    }
    fast += 1;
  }
  return slow;
}
